tianshou/policy/multiagent/mapolicy.py

```diff
@@ -0,0 +1,159 @@
+            actions.append(random.choice(legal_actions))
+        act = to_numpy(actions)
+        return Batch(act=act)
+
+    def learn(self, batch: Batch, **kwargs)\
+            -> Dict[str, Union[float, List[float]]]:
+        return {}
+
+
+class MultiAgentPolicyManager(BaseMultiAgentPolicy):
+    def __init__(self, policies: List[BaseMultiAgentPolicy]):
+        super().__init__()
+        self.policies = policies
+        for i, policy in enumerate(policies):
+            # agent_id 0 is reserved for the environment proxy
+            # (this MultiAgentPolicyManager)
+            policy.set_agent_id(i + 1)
+
+    def replace_policy(self, policy, agent_id):
+        self.policies[agent_id - 1] = policy
+        policy.set_agent_id(agent_id)
+
+    def forward(self,
+                batch: Batch,
+                state: Optional[Union[dict, Batch]] = None,
+                **kwargs) -> Batch:
+        """
+        :param state: if None, it means all agents have no state.
+            If not None, it should contain keys of agent_1, agent_2, ...
+
+        :return a Batch with the following contents
+        {
+        "act": actions corresponding to the input
+        "state":{
+            "agent_1": output state of agent_1's policy for the state
+            "agent_2": xxx
+            ...
+            "agent_n": xxx}
+        "out":{
+            "agent_1": output of agent_1's policy for the input
+            "agent_2": xxx
+            ...
+            "agent_n": xxx}
+        }
+        """
+        results = []
+        for policy in self.policies:
+            # This part of code is difficult to understand.
+            # Let's follow an example with two agents
+            # batch.obs.agent_id is [1, 2, 1, 2, 1, 2]
+            # each agent plays for three transitions
+            # agent_index for agent 1 is [0, 2, 4]
+            # agent_index for agent 2 is [1, 3, 5]
+            # we separate the transition of each agent
+            agent_index = np.nonzero(batch.obs.agent_id == policy.agent_id)[0]
+            if len(agent_index) == 0:
+                # (has_data, agent_index, out, act, state)
+                results.append((False, None, Batch(), None, Batch()))
+                continue
+            tmp_batch = batch[agent_index]
+            if isinstance(tmp_batch.rew, np.ndarray):
+                # reward can be empty Batch (after initial reset) or nparray.
+                tmp_batch.rew = tmp_batch.rew[:, policy.agent_id - 1]
+            out = policy(
+                batch=tmp_batch,
+                state=state and state["agent_" + str(policy.agent_id)],
+                **kwargs)
+            act = out.act
+            each_state = out.state \
+                if (hasattr(out, 'state') and out.state is not None) \
+                else Batch()
+            results.append((True, agent_index, out, act, each_state))
+        holder = Batch.cat([{'act': e[3]} for e in results if e[0]])
```

Comment:
Replace e by the actual content:
```
holder = Batch.cat([{'act': out} for (has_data, agent_index, out, act, each_state) in results if has_data])
```

It is longer but much clearer.

Reply:
Fixed in 4057231 .